onnxruntime/core/providers/cpu/sequence/sequence_ops.cc

```diff
@@ -96,9 +93,228 @@ Status SequenceAt::Compute(OpKernelContext* context) const {
+    SequenceEmpty);
+
+SequenceEmpty::SequenceEmpty(const OpKernelInfo& info) : OpKernel(info) {
+  if (!info.GetAttr("dtype", &dtype_).IsOK()) {
+    dtype_ = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+  }
+}
+
+Status SequenceEmpty::Compute(OpKernelContext* context) const {
+  auto* Y = context->Output<TensorSeq>(0);
+  ORT_ENFORCE(Y != nullptr, "SequenceEmpty: Got nullptr for output sequence");
+  MLDataType seq_dtype{};
+  switch (dtype_) {
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+      seq_dtype = DataTypeImpl::GetType<float>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_BOOL:
+      seq_dtype = DataTypeImpl::GetType<bool>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_INT32:
+      seq_dtype = DataTypeImpl::GetType<int>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:
+      seq_dtype = DataTypeImpl::GetType<double>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_STRING:
+      seq_dtype = DataTypeImpl::GetType<std::string>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_INT8:
+      seq_dtype = DataTypeImpl::GetType<int8_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT8:
+      seq_dtype = DataTypeImpl::GetType<uint8_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT16:
+      seq_dtype = DataTypeImpl::GetType<uint16_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_INT16:
+      seq_dtype = DataTypeImpl::GetType<int16_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64:
+      seq_dtype = DataTypeImpl::GetType<int64_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT32:
+      seq_dtype = DataTypeImpl::GetType<uint32_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT64:
+      seq_dtype = DataTypeImpl::GetType<uint64_t>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:
+      seq_dtype = DataTypeImpl::GetType<MLFloat16>();
+      break;
+    case ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16:
+      seq_dtype = DataTypeImpl::GetType<BFloat16>();
+      break;
+    default:
+      ORT_THROW("Unsupported 'dtype' value: ", dtype_);
+  }
+
+  Y->dtype = seq_dtype;
+  return Status::OK();
+}
+
+// SequenceInsert
+ONNX_CPU_OPERATOR_KERNEL(
+    SequenceInsert,
+    11,
+    KernelDefBuilder()
+        .TypeConstraint("S", DataTypeImpl::AllSequenceTensorTypes())
+        .TypeConstraint("I", std::vector<MLDataType>{
+                                 DataTypeImpl::GetTensorType<int32_t>(),
+                                 DataTypeImpl::GetTensorType<int64_t>()}),
+    SequenceInsert);
+
+Status CreateAndCopyCpuTensor(const Tensor& in_tensor, OpKernelContext* context, Tensor& out_tensor) {
+  AllocatorPtr alloc;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
+  out_tensor = Tensor(in_tensor.DataType(), onnxruntime::TensorShape(in_tensor.Shape()), alloc);
```

Comment:
Should this be using the temp space allocator?

Reply:
This eventually uses the allocator from the execution provider using device 0 and OrtMemTypeDefault (session_state.GetExecutionProviders().GetAllocator(memory_info);). The other alternative is to use OrtDefaultAllocator which default to CPU.